compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirLocalVariableAssignmentAnalyzer.kt

```diff
@@ -0,0 +1,349 @@
+         *
+         * Also, for simplicity, all conditionals are treated as non-exhaustive. Hence, a fallback edge is always added along a conditional
+         * structure.
+         *
+         * While building the mini CFG, inside each node, we collect local variables that are assigned later in the execution path.
+         *
+         * For example, consider the following code.
+         *
+         * ```
+         * fun test() {
+         *   var x: Int = 0
+         *   if (true) {
+         *     run {
+         *       x = 1
+         *       var y = 1
+         *       y = 2
+         *     }
+         *   } else {
+         *     x = 2
+         *   }
+         * }
+         * ```
+         *
+         * The generated mini CFG looks like
+         *
+         * ┌───────┐
+         * │  if   │
+         * └─┬─┬─┬─┘
+         *   │ │ │ fallback
+         *   │ │ └───────────────────────────┐
+         *   │ │ false                       │
+         *   │ └───────────────────────┐     │
+         *   │ true                    │     │
+         * ┌─┴─────┐               ┌───┴───┐ │
+         * │  run  │               │ empty │ │
+         * └─┬───┬─┘               └───┬───┘ │
+         *   │   │ normal execution    │     │
+         *   │   └─────────────┐       │     │
+         *   │ normal lambda   │       │     │
+         * ┌─┴──────┐      ┌───┴───┐   │     │
+         * │ lambda │      │ empty │   │     │
+         * │  body  │      │       │   │     │
+         * └────────┘      └───┬───┘   │     │
+         *   ┌─────────────────┘       │     │
+         *   │ ┌───────────────────────┘     │
+         *   │ │ ┌───────────────────────────┘
+         * ┌─┴─┴─┴─┐
+         * │ after │
+         * │  if   │
+         * └───────┘
+         *
+         * By the way, since local variables are not resolved at this point, we manually track local variable declarations and resolve them along
+         * the way so that shadowed names are handled correctly.
+         */
+        fun analyzeFunction(rootFunction: FirFunction<*>): FirLocalVariableAssignmentAnalyzer {
+            return FirLocalVariableAssignmentAnalyzer(computeAssignedLocalVariables(rootFunction))
+        }
+
+        /** Computes a mini CFG and returns the starting flow. */
+        private fun computeAssignedLocalVariables(firElement: FirElement): Map<FirFunction<*>, AssignedLocalVariables> {
+            val startFlow = MiniFlow.start()
+            val data = MiniCfgBuilder.MiniCfgData(startFlow)
+            MiniCfgBuilder().visitElement(firElement, data)
+            return data.localFunctionToAssignedLocalVariables
+        }
+
+        class AssignedLocalVariables(val outsideLocalFunction: Set<FirProperty>, val insideLocalFunction: Set<FirProperty>) {
+            val all get() = outsideLocalFunction + insideLocalFunction
+        }
+
+        private class MiniFlow(val parents: Set<MiniFlow>) {
```

Comment:
Btw, we need a better name. :) But nothing came to me now...

Reply:
Initialially I called it `ExecutionPathSegment`. Let me know if you prefer that or have some other suggestions.